crates/red_knot_python_semantic/src/types.rs

```diff
@@ -245,46 +263,77 @@ fn definition_expression_ty<'db>(
 fn bindings_ty<'db>(
     db: &'db dyn Db,
     bindings_with_constraints: BindingWithConstraintsIterator<'_, 'db>,
-) -> Option<Type<'db>> {
-    let mut def_types = bindings_with_constraints.map(
-        |BindingWithConstraints {
-             binding,
-             constraints,
-         }| {
-            let mut constraint_tys = constraints
-                .filter_map(|constraint| narrowing_constraint(db, constraint, binding))
-                .peekable();
-
-            let binding_ty = binding_ty(db, binding);
-            if constraint_tys.peek().is_some() {
-                constraint_tys
-                    .fold(
-                        IntersectionBuilder::new(db).add_positive(binding_ty),
-                        IntersectionBuilder::add_positive,
-                    )
-                    .build()
-            } else {
-                binding_ty
+) -> Symbol<'db> {
+    let mut is_unbound_visible = false;
+    let mut types = vec![];
+
+    for BindingWithConstraints {
+        binding,
+        constraints,
+        all_constraints,
+        visibility_constraints,
+        visibility_constraint,
+    } in bindings_with_constraints
+    {
+        let static_visibility =
+            visibility_constraints.analyze(db, all_constraints, visibility_constraint);
+
+        let Some(binding) = binding else {
+            if !static_visibility.is_always_false() {
+                // TODO: also distinguish between always-true and ambiguous?
```

Comment:
This seems like it would be more of a debug assertion / parity check than a behavior change, since "ambiguous" should always mean "there is at least one other visible binding" and "always-true" should mean "there are no other visible bindings".